File: benchmarks/rust/03_matrix_mul.rs
// Matrix Multiplication Benchmark - Rust Implementation
// Multiply two NxN matrices (matching Viper algorithm)

fn main() {
    let n = 30;
    let mut checksum: i64 = 0;
    
    for row in 0..n {
        for col in 0..n {
            let mut sum = 0;
            for k in 0..n {
                let a_val = (row * n + k) % 10;
                let b_val = (k * n + col + 1) % 10;
                sum += a_val * b_val;
            }
            checksum += sum as i64;
        }
    }
    
    println!("matrix {}x{} checksum: {}", n, n, checksum);
}

File: benchmarks/python/03_matrix_mul.py
def main():
    n = 30
    checksum = 0
    row = 0
    while row < n:
        col = 0
        while col < n:
            sum_val = 0
            k = 0
            while k < n:
                a_val = (row * n + k) % 10
                b_val = (k * n + col + 1) % 10
                sum_val += a_val * b_val
                k += 1
            checksum += sum_val
            col += 1
        row += 1
    print(checksum)


if __name__ == "__main__":
    main()
